File: Sources/UsageBarCore/UsageSnapshotV1.swift
import Foundation

/// Stable provider identifiers used by the public v1 telemetry contract.
public enum UsageProviderIDV1: String, Codable, CaseIterable, Sendable {
    case codex
    case claude
}

/// Provider collection state represented independently from UsageBar UI state.
public enum ProviderStateV1: String, Codable, Sendable {
    case disabled
    case unavailable
    case fresh
    case stale
}

/// Stable semantic kinds for normalized quota windows.
public enum UsageWindowKindV1: String, Codable, Sendable {
    case fiveHour
    case weekly
    case duration
    case unknown
}

/// Small, allowlisted public failure categories. Associated internal error text
/// is deliberately discarded before this boundary.
public enum ProviderErrorCodeV1: String, Codable, Sendable {
    case noData = "no_data"
    case notFound = "not_found"
    case untrustedExecutable = "untrusted_executable"
    case notAuthenticated = "not_authenticated"
    case timedOut = "timed_out"
    case unreadable
    case incompatible
    case commandFailed = "command_failed"
    case launchFailed = "launch_failed"
    case outputTooLarge = "output_too_large"
}

public struct ProviderErrorV1: Codable, Equatable, Sendable {
    public let code: ProviderErrorCodeV1

    public init(code: ProviderErrorCodeV1) {
        self.code = code
    }
}

public enum UsageSnapshotV1ValidationError: Error, Equatable {
    case unsupportedSchemaVersion(Int)
    case invalidProviderCardinality(UsageProviderIDV1, count: Int)
    case invalidProviderState(UsageProviderIDV1, ProviderStateV1)
    case invalidUsedPercent(Int)
    case invalidDurationMinutes(Int)
}

public struct UsageWindowV1: Codable, Equatable, Sendable {
    public let kind: UsageWindowKindV1
    public let durationMinutes: Int?
    public let usedPercent: Int
    public let resetAt: Date?

    public init(
        kind: UsageWindowKindV1,
        durationMinutes: Int?,
        usedPercent: Int,
        resetAt: Date?
    ) throws {
        guard (0...100).contains(usedPercent) else {
            throw UsageSnapshotV1ValidationError.invalidUsedPercent(usedPercent)
        }
        if let durationMinutes, durationMinutes <= 0 {
            throw UsageSnapshotV1ValidationError.invalidDurationMinutes(durationMinutes)
        }
        if kind == .duration, durationMinutes == nil {
            throw UsageSnapshotV1ValidationError.invalidDurationMinutes(0)
        }

        self.kind = kind
        self.durationMinutes = durationMinutes
        self.usedPercent = usedPercent
        self.resetAt = resetAt
    }

    private enum CodingKeys: String, CodingKey {
        case kind
        case durationMinutes
        case usedPercent
        case resetAt
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        guard container.contains(.durationMinutes) else {
            throw DecodingError.keyNotFound(
                CodingKeys.durationMinutes,
                .init(codingPath: container.codingPath, debugDescription: "Required nullable field is missing.")
            )
        }
        guard container.contains(.resetAt) else {
            throw DecodingError.keyNotFound(
                CodingKeys.resetAt,
                .init(codingPath: container.codingPath, debugDescription: "Required nullable field is missing.")
            )
        }
        try self.init(
            kind: container.decode(UsageWindowKindV1.self, forKey: .kind),
            durationMinutes: container.decodeIfPresent(Int.self, forKey: .durationMinutes),
            usedPercent: container.decode(Int.self, forKey: .usedPercent),
            resetAt: container.decodeIfPresent(Date.self, forKey: .resetAt)
        )
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(kind, forKey: .kind)
        try container.encode(durationMinutes, forKey: .durationMinutes)
        try container.encode(usedPercent, forKey: .usedPercent)
        try container.encode(resetAt, forKey: .resetAt)
    }
}

public struct ProviderSnapshotV1: Codable, Equatable, Sendable {
    public let id: UsageProviderIDV1
    public let state: ProviderStateV1
    public let lastSuccessfulAt: Date?
    public let error: ProviderErrorV1?
    public let windows: [UsageWindowV1]

    public init(
        id: UsageProviderIDV1,
        state: ProviderStateV1,
        lastSuccessfulAt: Date?,
        error: ProviderErrorV1?,
        windows: [UsageWindowV1]
    ) throws {
        let isValid: Bool
        switch state {
        case .disabled:
            isValid = lastSuccessfulAt == nil && error == nil && windows.isEmpty
        case .unavailable:
            isValid = lastSuccessfulAt == nil && error != nil && windows.isEmpty
        case .fresh:
            isValid = lastSuccessfulAt != nil && error == nil && !windows.isEmpty
        case .stale:
            isValid = lastSuccessfulAt != nil && error != nil && !windows.isEmpty
        }
        guard isValid else {
            throw UsageSnapshotV1ValidationError.invalidProviderState(id, state)
        }

        self.id = id
        self.state = state
        self.lastSuccessfulAt = lastSuccessfulAt
        self.error = error
        self.windows = windows
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case state
        case lastSuccessfulAt
        case error
        case windows
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        guard container.contains(.lastSuccessfulAt) else {
            throw DecodingError.keyNotFound(
                CodingKeys.lastSuccessfulAt,
                .init(codingPath: container.codingPath, debugDescription: "Required nullable field is missing.")
            )
        }
        guard container.contains(.error) else {
            throw DecodingError.keyNotFound(
                CodingKeys.error,
                .init(codingPath: container.codingPath, debugDescription: "Required nullable field is missing.")
            )
        }
        try self.init(
            id: container.decode(UsageProviderIDV1.self, forKey: .id),
            state: container.decode(ProviderStateV1.self, forKey: .state),
            lastSuccessfulAt: container.decodeIfPresent(Date.self, forKey: .lastSuccessfulAt),
            error: container.decodeIfPresent(ProviderErrorV1.self, forKey: .error),
            windows: container.decode([UsageWindowV1].self, forKey: .windows)
        )
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(state, forKey: .state)
        try container.encode(lastSuccessfulAt, forKey: .lastSuccessfulAt)
        try container.encode(error, forKey: .error)
        try container.encode(windows, forKey: .windows)
    }
}

public struct UsageSnapshotV1: Codable, Equatable, Sendable {
    public static let currentSchemaVersion = 1

    public let schemaVersion: Int
    public let observedAt: Date
    public let providers: [ProviderSnapshotV1]

    public init(observedAt: Date, providers: [ProviderSnapshotV1]) throws {
        self.schemaVersion = Self.currentSchemaVersion
        self.observedAt = observedAt
        self.providers = try Self.validatedProviders(providers)
    }

    private init(schemaVersion: Int, observedAt: Date, providers: [ProviderSnapshotV1]) throws {
        guard schemaVersion == Self.currentSchemaVersion else {
            throw UsageSnapshotV1ValidationError.unsupportedSchemaVersion(schemaVersion)
        }
        self.schemaVersion = schemaVersion
        self.observedAt = observedAt
        self.providers = try Self.validatedProviders(providers)
    }

    private static func validatedProviders(_ providers: [ProviderSnapshotV1]) throws -> [ProviderSnapshotV1] {
        for id in UsageProviderIDV1.allCases {
            let count = providers.filter { $0.id == id }.count
            guard count == 1 else {
                throw UsageSnapshotV1ValidationError.invalidProviderCardinality(id, count: count)
            }
        }
        return UsageProviderIDV1.allCases.compactMap { id in
            providers.first { $0.id == id }
        }
    }

    private enum CodingKeys: String, CodingKey {
        case schemaVersion
        case observedAt
        case providers
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        try self.init(
            schemaVersion: container.decode(Int.self, forKey: .schemaVersion),
            observedAt: container.decode(Date.self, forKey: .observedAt),
            providers: container.decode([ProviderSnapshotV1].self, forKey: .providers)
        )
    }
}

/// Pure input boundary for projecting the current accepted UsageBar state.
/// Connection flags are separate because disabled providers are absent from the
/// existing in-memory provider dictionary.
public struct UsageSnapshotV1ProjectionInput {
    public let codexIsEnabled: Bool
    public let claudeIsEnabled: Bool
    public let usages: [String: ProviderUsage]

    public init(codexIsEnabled: Bool, claudeIsEnabled: Bool, usages: [String: ProviderUsage]) {
        self.codexIsEnabled = codexIsEnabled
        self.claudeIsEnabled = claudeIsEnabled
        self.usages = usages
    }
}

public enum UsageSnapshotV1Projection {
    public static func project(
        _ input: UsageSnapshotV1ProjectionInput,
        observedAt: Date
    ) throws -> UsageSnapshotV1 {
        let providers = try [
            provider(
                id: .codex,
                isEnabled: input.codexIsEnabled,
                usage: input.usages["Codex"]
            ),
            provider(
                id: .claude,
                isEnabled: input.claudeIsEnabled,
                usage: input.usages["Claude Code"]
            )
        ]
        return try UsageSnapshotV1(observedAt: observedAt, providers: providers)
    }

    private static func provider(
        id: UsageProviderIDV1,
        isEnabled: Bool,
        usage: ProviderUsage?
    ) throws -> ProviderSnapshotV1 {
        guard isEnabled else {
            return try ProviderSnapshotV1(
                id: id,
                state: .disabled,
                lastSuccessfulAt: nil,
                error: nil,
                windows: []
            )
        }

        guard let usage else {
            return try unavailable(id: id, issue: nil)
        }

        guard !usage.windows.isEmpty else {
            guard usage.lastSuccessfulAt == nil else {
                throw UsageSnapshotV1ValidationError.invalidProviderState(id, .unavailable)
            }
            return try unavailable(id: id, issue: usage.error)
        }

        guard let lastSuccessfulAt = usage.lastSuccessfulAt else {
            throw UsageSnapshotV1ValidationError.invalidProviderState(
                id,
                usage.error == nil ? .fresh : .stale
            )
        }

        let windows = try usage.windows.map { window in
            try UsageWindowV1(
                kind: kind(for: window.kind),
                durationMinutes: window.durationMinutes,
                usedPercent: window.usedPercent,
                resetAt: window.resetsAt
            )
        }
        let publicError = usage.error.map { ProviderErrorV1(code: errorCode(for: $0)) }
        return try ProviderSnapshotV1(
            id: id,
            state: publicError == nil ? .fresh : .stale,
            lastSuccessfulAt: lastSuccessfulAt,
            error: publicError,
            windows: windows
        )
    }

    private static func unavailable(
        id: UsageProviderIDV1,
        issue: ProviderIssue?
    ) throws -> ProviderSnapshotV1 {
        let code = issue.map(errorCode(for:)) ?? .noData
        return try ProviderSnapshotV1(
            id: id,
            state: .unavailable,
            lastSuccessfulAt: nil,
            error: ProviderErrorV1(code: code),
            windows: []
        )
    }

    private static func kind(for kind: UsageWindowKind) -> UsageWindowKindV1 {
        switch kind {
        case .fiveHour: return .fiveHour
        case .weekly: return .weekly
        case .duration: return .duration
        case .unknown: return .unknown
        }
    }

    private static func errorCode(for issue: ProviderIssue) -> ProviderErrorCodeV1 {
        switch issue {
        case .refreshing, .noData:
            return .noData
        case .codexNotFound, .claudeNotFound:
            return .notFound
        case .codexUntrustedExecutable, .claudeUntrustedExecutable:
            return .untrustedExecutable
        case .claudeNotLoggedIn:
            return .notAuthenticated
        case .codexTimedOut, .claudeUsageTimedOut:
            return .timedOut
        case .codexUsageUnavailable, .codexLimitMissing, .codexEmptyResponse, .claudeUsageUnreadable:
            return .unreadable
        case .codexIncompatible:
            return .incompatible
        case .codexCommandFailed:
            return .commandFailed
        case .codexLaunchFailed, .claudeLaunchFailed:
            return .launchFailed
        case .outputTooLarge:
            return .outputTooLarge
        }
    }
}

/// Canonical v1 JSON codec. The wire format uses sorted object keys and UTC
/// RFC 3339 timestamps with millisecond precision.
public enum UsageSnapshotV1JSON {
    public static func encode(_ snapshot: UsageSnapshotV1) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        encoder.dateEncodingStrategy = .custom { date, encoder in
            var container = encoder.singleValueContainer()
            try container.encode(timestampString(from: date))
        }
        return try encoder.encode(snapshot)
    }

    public static func decode(_ data: Data) throws -> UsageSnapshotV1 {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let value = try container.decode(String.self)
            guard value.hasSuffix("Z"), let date = timestampDate(from: value) else {
                throw DecodingError.dataCorruptedError(
                    in: container,
                    debugDescription: "Expected an RFC 3339 UTC timestamp with millisecond precision."
                )
            }
            return date
        }
        return try decoder.decode(UsageSnapshotV1.self, from: data)
    }

    private static func timestampString(from date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: date)
    }

    private static func timestampDate(from value: String) -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.date(from: value)
    }
}

File: windows/src/UsageBar.Windows.Core/Contract/UsageSnapshotV1.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;
using UsageBar.Windows.Core.Providers;
using UsageBar.Windows.Core.Settings;

namespace UsageBar.Windows.Core.Contract;

/// <summary>Stable public provider identifiers for the v1 telemetry protocol.</summary>
[JsonConverter(typeof(UsageProviderIdV1JsonConverter))]
public enum UsageProviderIdV1
{
    Codex,
    Claude
}

/// <summary>Provider collection state, independent of presentation state.</summary>
[JsonConverter(typeof(ProviderStateV1JsonConverter))]
public enum ProviderStateV1
{
    Disabled,
    Unavailable,
    Fresh,
    Stale
}

/// <summary>Stable semantic kinds for normalized quota windows.</summary>
[JsonConverter(typeof(UsageWindowKindV1JsonConverter))]
public enum UsageWindowKindV1
{
    FiveHour,
    Weekly,
    Duration,
    Unknown
}

/// <summary>
/// Small, privacy-safe public failure taxonomy. Internal error details never
/// cross this boundary.
/// </summary>
[JsonConverter(typeof(ProviderErrorCodeV1JsonConverter))]
public enum ProviderErrorCodeV1
{
    NoData,
    NotFound,
    UntrustedExecutable,
    NotAuthenticated,
    TimedOut,
    Unreadable,
    Incompatible,
    CommandFailed,
    LaunchFailed,
    OutputTooLarge
}

/// <summary>Explicit failure from v1 projection or wire validation.</summary>
public sealed class UsageSnapshotV1ValidationException : Exception
{
    public UsageSnapshotV1ValidationException(string reason)
        : base($"Invalid UsageSnapshotV1: {reason}.")
    {
        Reason = reason;
    }

    public string Reason { get; }
}

public sealed record ProviderErrorV1
{
    [JsonConstructor]
    public ProviderErrorV1(ProviderErrorCodeV1 code)
    {
        Code = code;
    }

    [JsonPropertyName("code")]
    [JsonPropertyOrder(0)]
    public ProviderErrorCodeV1 Code { get; }
}

public sealed record UsageWindowV1
{
    [JsonConstructor]
    public UsageWindowV1(
        UsageWindowKindV1 kind,
        int? durationMinutes,
        int usedPercent,
        DateTimeOffset? resetAt)
    {
        if (!Enum.IsDefined(kind))
        {
            throw new UsageSnapshotV1ValidationException("unsupported_window_kind");
        }

        if (usedPercent is < 0 or > 100)
        {
            throw new UsageSnapshotV1ValidationException("invalid_used_percent");
        }

        if (durationMinutes is <= 0)
        {
            throw new UsageSnapshotV1ValidationException("invalid_duration_minutes");
        }

        if (kind == UsageWindowKindV1.Duration && durationMinutes is null)
        {
            throw new UsageSnapshotV1ValidationException("missing_duration_minutes");
        }

        Kind = kind;
        DurationMinutes = durationMinutes;
        UsedPercent = usedPercent;
        ResetAt = resetAt;
    }

    [JsonPropertyName("kind")]
    [JsonPropertyOrder(0)]
    public UsageWindowKindV1 Kind { get; }

    [JsonPropertyName("durationMinutes")]
    [JsonPropertyOrder(1)]
    public int? DurationMinutes { get; }

    [JsonPropertyName("usedPercent")]
    [JsonPropertyOrder(2)]
    public int UsedPercent { get; }

    [JsonPropertyName("resetAt")]
    [JsonPropertyOrder(3)]
    public DateTimeOffset? ResetAt { get; }
}

public sealed record ProviderSnapshotV1
{
    [JsonConstructor]
    public ProviderSnapshotV1(
        UsageProviderIdV1 id,
        ProviderStateV1 state,
        DateTimeOffset? lastSuccessfulAt,
        ProviderErrorV1? error,
        IReadOnlyList<UsageWindowV1> windows)
    {
        ArgumentNullException.ThrowIfNull(windows);

        if (!Enum.IsDefined(id) || !Enum.IsDefined(state))
        {
            throw new UsageSnapshotV1ValidationException("unsupported_provider_value");
        }

        var isValid = state switch
        {
            ProviderStateV1.Disabled =>
                lastSuccessfulAt is null && error is null && windows.Count == 0,
            ProviderStateV1.Unavailable =>
                lastSuccessfulAt is null && error is not null && windows.Count == 0,
            ProviderStateV1.Fresh =>
                lastSuccessfulAt is not null && error is null && windows.Count > 0,
            ProviderStateV1.Stale =>
                lastSuccessfulAt is not null && error is not null && windows.Count > 0,
            _ => false
        };
        if (!isValid)
        {
            throw new UsageSnapshotV1ValidationException("invalid_provider_state");
        }

        Id = id;
        State = state;
        LastSuccessfulAt = lastSuccessfulAt;
        Error = error;
        Windows = windows.ToArray();
    }

    [JsonPropertyName("id")]
    [JsonPropertyOrder(0)]
    public UsageProviderIdV1 Id { get; }

    [JsonPropertyName("state")]
    [JsonPropertyOrder(1)]
    public ProviderStateV1 State { get; }

    [JsonPropertyName("lastSuccessfulAt")]
    [JsonPropertyOrder(2)]
    public DateTimeOffset? LastSuccessfulAt { get; }

    [JsonPropertyName("error")]
    [JsonPropertyOrder(3)]
    public ProviderErrorV1? Error { get; }

    [JsonPropertyName("windows")]
    [JsonPropertyOrder(4)]
    public IReadOnlyList<UsageWindowV1> Windows { get; }
}

public sealed record UsageSnapshotV1
{
    public const int CurrentSchemaVersion = 1;

    public UsageSnapshotV1(
        DateTimeOffset observedAt,
        IReadOnlyList<ProviderSnapshotV1> providers)
        : this(CurrentSchemaVersion, observedAt, providers)
    {
    }

    [JsonConstructor]
    public UsageSnapshotV1(
        int schemaVersion,
        DateTimeOffset observedAt,
        IReadOnlyList<ProviderSnapshotV1> providers)
    {
        ArgumentNullException.ThrowIfNull(providers);

        if (schemaVersion != CurrentSchemaVersion)
        {
            throw new UsageSnapshotV1ValidationException("unsupported_schema_version");
        }

        foreach (var id in new[] { UsageProviderIdV1.Codex, UsageProviderIdV1.Claude })
        {
            if (providers.Count(provider => provider.Id == id) != 1)
            {
                throw new UsageSnapshotV1ValidationException("invalid_provider_cardinality");
            }
        }

        SchemaVersion = schemaVersion;
        ObservedAt = observedAt;
        Providers = new[]
        {
            providers.Single(provider => provider.Id == UsageProviderIdV1.Codex),
            providers.Single(provider => provider.Id == UsageProviderIdV1.Claude)
        };
    }

    [JsonPropertyName("schemaVersion")]
    [JsonPropertyOrder(0)]
    public int SchemaVersion { get; }

    [JsonPropertyName("observedAt")]
    [JsonPropertyOrder(1)]
    public DateTimeOffset ObservedAt { get; }

    [JsonPropertyName("providers")]
    [JsonPropertyOrder(2)]
    public IReadOnlyList<ProviderSnapshotV1> Providers { get; }
}

/// <summary>
/// Pure projection input. Connection flags remain separate because disabled
/// providers are absent from the existing in-memory usage dictionary.
/// </summary>
public sealed record UsageSnapshotV1ProjectionInput
{
    public UsageSnapshotV1ProjectionInput(
        bool codexIsEnabled,
        bool claudeIsEnabled,
        IReadOnlyDictionary<string, ProviderUsage> usages)
    {
        ArgumentNullException.ThrowIfNull(usages);
        CodexIsEnabled = codexIsEnabled;
        ClaudeIsEnabled = claudeIsEnabled;
        Usages = usages;
    }

    public bool CodexIsEnabled { get; }

    public bool ClaudeIsEnabled { get; }

    public IReadOnlyDictionary<string, ProviderUsage> Usages { get; }

    public static UsageSnapshotV1ProjectionInput FromSettings(
        UsageBarSettings settings,
        IReadOnlyDictionary<string, ProviderUsage> usages)
    {
        ArgumentNullException.ThrowIfNull(settings);
        return new(settings.CodexConnected, settings.ClaudeConnected, usages);
    }
}

/// <summary>
/// Deterministically projects accepted raw Windows Core telemetry. It never
/// collects provider data, reads UI state, persists data, or consults a clock.
/// </summary>
public static class UsageSnapshotV1Projection
{
    public static UsageSnapshotV1 Project(
        UsageSnapshotV1ProjectionInput input,
        DateTimeOffset observedAt)
    {
        ArgumentNullException.ThrowIfNull(input);

        input.Usages.TryGetValue(ProviderNames.Codex, out var codex);
        input.Usages.TryGetValue(ProviderNames.ClaudeCode, out var claude);

        return new UsageSnapshotV1(
            observedAt,
            new[]
            {
                Provider(UsageProviderIdV1.Codex, input.CodexIsEnabled, codex),
                Provider(UsageProviderIdV1.Claude, input.ClaudeIsEnabled, claude)
            });
    }

    private static ProviderSnapshotV1 Provider(
        UsageProviderIdV1 id,
        bool isEnabled,
        ProviderUsage? usage)
    {
        if (!isEnabled)
        {
            return new ProviderSnapshotV1(
                id,
                ProviderStateV1.Disabled,
                lastSuccessfulAt: null,
                error: null,
                Array.Empty<UsageWindowV1>());
        }

        if (usage is null)
        {
            return Unavailable(id, issue: null);
        }

        if (usage.Windows.Count == 0)
        {
            if (usage.LastSuccessfulAt is not null)
            {
                throw new UsageSnapshotV1ValidationException("unavailable_with_last_success");
            }

            return Unavailable(id, usage.Error);
        }

        if (usage.LastSuccessfulAt is not { } lastSuccessfulAt)
        {
            throw new UsageSnapshotV1ValidationException("windows_without_last_success");
        }

        var windows = usage.Windows.Select(Window).ToArray();
        var error = usage.Error is null
            ? null
            : new ProviderErrorV1(ErrorCode(usage.Error));
        return new ProviderSnapshotV1(
            id,
            error is null ? ProviderStateV1.Fresh : ProviderStateV1.Stale,
            lastSuccessfulAt,
            error,
            windows);
    }

    private static ProviderSnapshotV1 Unavailable(
        UsageProviderIdV1 id,
        ProviderIssue? issue)
    {
        var code = issue is null ? ProviderErrorCodeV1.NoData : ErrorCode(issue);
        return new ProviderSnapshotV1(
            id,
            ProviderStateV1.Unavailable,
            lastSuccessfulAt: null,
            new ProviderErrorV1(code),
            Array.Empty<UsageWindowV1>());
    }

    private static UsageWindowV1 Window(UsageWindow window)
    {
        ArgumentNullException.ThrowIfNull(window);
        return new UsageWindowV1(
            window.Kind.CategoryKind switch
            {
                UsageWindowKind.Category.FiveHour => UsageWindowKindV1.FiveHour,
                UsageWindowKind.Category.Weekly => UsageWindowKindV1.Weekly,
                UsageWindowKind.Category.Duration => UsageWindowKindV1.Duration,
                UsageWindowKind.Category.Unknown => UsageWindowKindV1.Unknown,
                _ => throw new UsageSnapshotV1ValidationException("unsupported_window_kind")
            },
            window.DurationMinutes,
            window.UsedPercent,
            window.ResetsAt);
    }

    internal static ProviderErrorCodeV1 ErrorCode(ProviderIssue issue)
    {
        ArgumentNullException.ThrowIfNull(issue);
        return issue.Code switch
        {
            ProviderIssueCode.Refreshing or ProviderIssueCode.NoData =>
                ProviderErrorCodeV1.NoData,
            ProviderIssueCode.CodexNotFound or ProviderIssueCode.ClaudeNotFound or
                ProviderIssueCode.ClaudeWslUnavailable or
                ProviderIssueCode.ClaudeWslDistributionUnavailable =>
                ProviderErrorCodeV1.NotFound,
            ProviderIssueCode.CodexUntrustedExecutable or
                ProviderIssueCode.ClaudeUntrustedExecutable =>
                ProviderErrorCodeV1.UntrustedExecutable,
            ProviderIssueCode.ClaudeNotLoggedIn =>
                ProviderErrorCodeV1.NotAuthenticated,
            ProviderIssueCode.CodexTimedOut or ProviderIssueCode.ClaudeUsageTimedOut =>
                ProviderErrorCodeV1.TimedOut,
            ProviderIssueCode.CodexUsageUnavailable or ProviderIssueCode.CodexLimitMissing or
                ProviderIssueCode.CodexEmptyResponse or ProviderIssueCode.ClaudeUsageUnreadable =>
                ProviderErrorCodeV1.Unreadable,
            ProviderIssueCode.CodexIncompatible or
                ProviderIssueCode.CodexUnsupportedInstallation or
                ProviderIssueCode.ClaudeUnsupportedInstallation or
                ProviderIssueCode.ClaudeGitBashMissing =>
                ProviderErrorCodeV1.Incompatible,
            ProviderIssueCode.CodexCommandFailed or ProviderIssueCode.ClaudeCommandFailed or
                ProviderIssueCode.Cancelled =>
                ProviderErrorCodeV1.CommandFailed,
            ProviderIssueCode.CodexLaunchFailed or ProviderIssueCode.ClaudeLaunchFailed =>
                ProviderErrorCodeV1.LaunchFailed,
            ProviderIssueCode.OutputTooLarge =>
                ProviderErrorCodeV1.OutputTooLarge,
            _ => throw new UsageSnapshotV1ValidationException("unsupported_provider_issue")
        };
    }
}

/// <summary>
/// Canonical v1 JSON codec. Nullable fields are always present and timestamps
/// are UTC RFC 3339 strings with exactly millisecond precision.
/// </summary>
public static class UsageSnapshotV1Json
{
    private static readonly JsonSerializerOptions Options = CreateOptions();

    public static byte[] Encode(UsageSnapshotV1 snapshot)
    {
        ArgumentNullException.ThrowIfNull(snapshot);
        return JsonSerializer.SerializeToUtf8Bytes(snapshot, Options);
    }

    public static UsageSnapshotV1 Decode(ReadOnlySpan<byte> data)
    {
        ValidateRequiredFields(data);
        var snapshot = JsonSerializer.Deserialize<UsageSnapshotV1>(data, Options);
        return snapshot ?? throw new JsonException("Expected a v1 usage snapshot object.");
    }

    public static UsageSnapshotV1 Decode(string json)
    {
        ArgumentNullException.ThrowIfNull(json);
        ValidateRequiredFields(json);
        var snapshot = JsonSerializer.Deserialize<UsageSnapshotV1>(json, Options);
        return snapshot ?? throw new JsonException("Expected a v1 usage snapshot object.");
    }

    private static void ValidateRequiredFields(ReadOnlySpan<byte> data)
    {
        using var document = JsonDocument.Parse(data.ToArray());
        ValidateRequiredFields(document.RootElement);
    }

    private static void ValidateRequiredFields(string json)
    {
        using var document = JsonDocument.Parse(json);
        ValidateRequiredFields(document.RootElement);
    }

    private static void ValidateRequiredFields(JsonElement root)
    {
        RequireObjectProperties(root, "schemaVersion", "observedAt", "providers");
        var providers = root.GetProperty("providers");
        if (providers.ValueKind != JsonValueKind.Array)
        {
            throw new JsonException("Expected providers to be an array.");
        }

        foreach (var provider in providers.EnumerateArray())
        {
            RequireObjectProperties(provider, "id", "state", "lastSuccessfulAt", "error", "windows");
            var error = provider.GetProperty("error");
            if (error.ValueKind != JsonValueKind.Null)
            {
                RequireObjectProperties(error, "code");
            }

            var windows = provider.GetProperty("windows");
            if (windows.ValueKind != JsonValueKind.Array)
            {
                throw new JsonException("Expected windows to be an array.");
            }

            foreach (var window in windows.EnumerateArray())
            {
                RequireObjectProperties(window, "kind", "durationMinutes", "usedPercent", "resetAt");
            }
        }
    }

    private static void RequireObjectProperties(JsonElement element, params string[] names)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            throw new JsonException("Expected a v1 contract object.");
        }

        foreach (var name in names)
        {
            if (!element.TryGetProperty(name, out _))
            {
                throw new JsonException($"Missing required v1 field: {name}.");
            }
        }
    }

    private static JsonSerializerOptions CreateOptions()
    {
        var options = new JsonSerializerOptions
        {
            DefaultIgnoreCondition = JsonIgnoreCondition.Never,
            PropertyNameCaseInsensitive = false,
            WriteIndented = false
        };
        options.Converters.Add(new CanonicalUtcDateTimeOffsetJsonConverter());
        return options;
    }
}

internal sealed class CanonicalUtcDateTimeOffsetJsonConverter : JsonConverter<DateTimeOffset>
{
    private const string Format = "yyyy-MM-dd'T'HH:mm:ss.fff'Z'";

    public override DateTimeOffset Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options)
    {
        var value = reader.GetString();
        if (value is null ||
            !DateTimeOffset.TryParseExact(
                value,
                Format,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                out var timestamp))
        {
            throw new JsonException("Expected an RFC 3339 UTC timestamp with millisecond precision.");
        }

        return timestamp;
    }

    public override void Write(
        Utf8JsonWriter writer,
        DateTimeOffset value,
        JsonSerializerOptions options) =>
        writer.WriteStringValue(value.ToUniversalTime().ToString(Format, CultureInfo.InvariantCulture));
}

internal sealed class UsageProviderIdV1JsonConverter : JsonConverter<UsageProviderIdV1>
{
    public override UsageProviderIdV1 Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options) => reader.GetString() switch
        {
            "codex" => UsageProviderIdV1.Codex,
            "claude" => UsageProviderIdV1.Claude,
            _ => throw new JsonException("Unknown v1 provider id.")
        };

    public override void Write(Utf8JsonWriter writer, UsageProviderIdV1 value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            UsageProviderIdV1.Codex => "codex",
            UsageProviderIdV1.Claude => "claude",
            _ => throw new JsonException("Unknown v1 provider id.")
        });
}

internal sealed class ProviderStateV1JsonConverter : JsonConverter<ProviderStateV1>
{
    public override ProviderStateV1 Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options) => reader.GetString() switch
        {
            "disabled" => ProviderStateV1.Disabled,
            "unavailable" => ProviderStateV1.Unavailable,
            "fresh" => ProviderStateV1.Fresh,
            "stale" => ProviderStateV1.Stale,
            _ => throw new JsonException("Unknown v1 provider state.")
        };

    public override void Write(Utf8JsonWriter writer, ProviderStateV1 value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            ProviderStateV1.Disabled => "disabled",
            ProviderStateV1.Unavailable => "unavailable",
            ProviderStateV1.Fresh => "fresh",
            ProviderStateV1.Stale => "stale",
            _ => throw new JsonException("Unknown v1 provider state.")
        });
}

internal sealed class UsageWindowKindV1JsonConverter : JsonConverter<UsageWindowKindV1>
{
    public override UsageWindowKindV1 Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options) => reader.GetString() switch
        {
            "fiveHour" => UsageWindowKindV1.FiveHour,
            "weekly" => UsageWindowKindV1.Weekly,
            "duration" => UsageWindowKindV1.Duration,
            "unknown" => UsageWindowKindV1.Unknown,
            _ => throw new JsonException("Unknown v1 window kind.")
        };

    public override void Write(Utf8JsonWriter writer, UsageWindowKindV1 value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            UsageWindowKindV1.FiveHour => "fiveHour",
            UsageWindowKindV1.Weekly => "weekly",
            UsageWindowKindV1.Duration => "duration",
            UsageWindowKindV1.Unknown => "unknown",
            _ => throw new JsonException("Unknown v1 window kind.")
        });
}

internal sealed class ProviderErrorCodeV1JsonConverter : JsonConverter<ProviderErrorCodeV1>
{
    public override ProviderErrorCodeV1 Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options) => reader.GetString() switch
        {
            "no_data" => ProviderErrorCodeV1.NoData,
            "not_found" => ProviderErrorCodeV1.NotFound,
            "untrusted_executable" => ProviderErrorCodeV1.UntrustedExecutable,
            "not_authenticated" => ProviderErrorCodeV1.NotAuthenticated,
            "timed_out" => ProviderErrorCodeV1.TimedOut,
            "unreadable" => ProviderErrorCodeV1.Unreadable,
            "incompatible" => ProviderErrorCodeV1.Incompatible,
            "command_failed" => ProviderErrorCodeV1.CommandFailed,
            "launch_failed" => ProviderErrorCodeV1.LaunchFailed,
            "output_too_large" => ProviderErrorCodeV1.OutputTooLarge,
            _ => throw new JsonException("Unknown v1 provider error code.")
        };

    public override void Write(Utf8JsonWriter writer, ProviderErrorCodeV1 value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            ProviderErrorCodeV1.NoData => "no_data",
            ProviderErrorCodeV1.NotFound => "not_found",
            ProviderErrorCodeV1.UntrustedExecutable => "untrusted_executable",
            ProviderErrorCodeV1.NotAuthenticated => "not_authenticated",
            ProviderErrorCodeV1.TimedOut => "timed_out",
            ProviderErrorCodeV1.Unreadable => "unreadable",
            ProviderErrorCodeV1.Incompatible => "incompatible",
            ProviderErrorCodeV1.CommandFailed => "command_failed",
            ProviderErrorCodeV1.LaunchFailed => "launch_failed",
            ProviderErrorCodeV1.OutputTooLarge => "output_too_large",
            _ => throw new JsonException("Unknown v1 provider error code.")
        });
}
